admittance_controller/src/admittance_controller.cpp

```diff
@@ -0,0 +1,505 @@
+    [this](const std::shared_ptr<trajectory_msgs::msg::JointTrajectoryPoint> msg)
+  { input_joint_command_.writeFromNonRT(msg); };
+  input_joint_command_subscriber_ =
+    get_node()->create_subscription<trajectory_msgs::msg::JointTrajectoryPoint>(
+      "~/joint_commands", rclcpp::SystemDefaultsQoS(), joint_command_callback);
+  s_publisher_ = get_node()->create_publisher<control_msgs::msg::AdmittanceControllerState>(
+    "~/state", rclcpp::SystemDefaultsQoS());
+  state_publisher_ =
+    std::make_unique<realtime_tools::RealtimePublisher<ControllerStateMsg>>(s_publisher_);
+
+  // Initialize state message
+  state_publisher_->lock();
+  state_publisher_->msg_ = admittance_->get_controller_state();
+  state_publisher_->unlock();
+
+  // Initialize FTS semantic semantic_component
+  force_torque_sensor_ = std::make_unique<semantic_components::ForceTorqueSensor>(
+    semantic_components::ForceTorqueSensor(admittance_->parameters_.ft_sensor.name));
+
+  // configure admittance rule
+  if (admittance_->configure(get_node(), num_joints_) == controller_interface::return_type::ERROR)
+  {
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  return LifecycleNodeInterface::on_configure(previous_state);
+}
+
+controller_interface::CallbackReturn AdmittanceController::on_activate(
+  const rclcpp_lifecycle::State & previous_state)
+{
+  // on_activate is called when the lifecycle node activates.
+  if (!admittance_)
+  {
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  // update parameters if any have changed
+  if (admittance_->parameter_handler_->is_old(admittance_->parameters_))
+  {
+    admittance_->parameters_ = admittance_->parameter_handler_->get_params();
+  }
+
+  // get state interface inds
+  std::unordered_map<std::string, size_t> inter_to_ind = {
+    {hardware_interface::HW_IF_POSITION, -1},
+    {hardware_interface::HW_IF_VELOCITY, -1},
+    {hardware_interface::HW_IF_ACCELERATION, -1}};
+  for (auto i = 0ul; i < admittance_->parameters_.state_interfaces.size(); i++)
+  {
+    const auto & interface = admittance_->parameters_.state_interfaces[i];
+    inter_to_ind[interface] = i;
+  }
+  state_pos_ind = inter_to_ind[hardware_interface::HW_IF_POSITION];
+  state_vel_ind = inter_to_ind[hardware_interface::HW_IF_VELOCITY];
+  state_acc_ind = inter_to_ind[hardware_interface::HW_IF_ACCELERATION];
+
+  // get state interface inds
+  inter_to_ind = {
+    {hardware_interface::HW_IF_POSITION, -1},
+    {hardware_interface::HW_IF_VELOCITY, -1},
+    {hardware_interface::HW_IF_ACCELERATION, -1}};
+  for (auto i = 0ul; i < admittance_->parameters_.command_interfaces.size(); i++)
+  {
+    const auto & interface = admittance_->parameters_.command_interfaces[i];
+    inter_to_ind[interface] = i;
+  }
+  command_pos_ind = inter_to_ind[hardware_interface::HW_IF_POSITION];
+  command_vel_ind = inter_to_ind[hardware_interface::HW_IF_VELOCITY];
+  command_acc_ind = inter_to_ind[hardware_interface::HW_IF_ACCELERATION];
```

Comment:
Why do we need this block? I assume it is for sorting interfaces. Why do you not use the same approach as in [JointTrajectoryController](https://github.com/ros-controls/ros2_controllers/blob/93fa9fec93321960756429830b661ca5d0b3b013/joint_trajectory_controller/src/joint_trajectory_controller.cpp#L813)?